File: packages/cli/src/commands/doctor.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * doctor command — diagnose Matimo setup and report actionable issues.
 * matimo doctor
 */

interface Issue {
  severity: 'error' | 'warn' | 'info';
  message: string;
}

const issues: Issue[] = [];

function check(
  label: string,
  pass: boolean,
  message: string,
  severity: 'error' | 'warn' = 'error'
): void {
  const icon = pass ? '✅' : severity === 'error' ? '❌' : '⚠️ ';
  console.info(`  ${icon} ${label}`);
  if (!pass) {
    issues.push({ severity, message });
    console.info(`     ${message}`);
  }
}

function findNodeModules(): string | null {
  let dir = process.cwd();
  for (let i = 0; i < 10; i++) {
    const candidate = path.join(dir, 'node_modules');
    if (fs.existsSync(candidate)) return candidate;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

async function loadMatimoCoreExport<T>(exportName: string): Promise<T | null> {
  try {
    // Dynamic import works in both ESM (built CLI) and CJS (ts-jest test transform)
    const core = (await import('@matimo/core')) as Record<string, unknown>;
    return (core[exportName] as T) ?? null;
  } catch {
    return null;
  }
}

export async function doctorCommand(): Promise<void> {
  // Ensure issues do not accumulate across multiple invocations
  issues.length = 0;
  console.info('\n🩺 Matimo Doctor — Checking your setup...\n');

  // ─── 1. Node.js version ────────────────────────────────────────────────
  console.info('Node.js:');
  const nodeVersion = process.versions.node;
  const [major] = nodeVersion.split('.').map(Number);
  check(
    `Node.js v${nodeVersion}`,
    major >= 18,
    `Node.js 18+ required. You are running ${nodeVersion}. Upgrade: https://nodejs.org`
  );
  console.info('');

  // ─── 2. @matimo/* packages ─────────────────────────────────────────────
  const nodeModulesPath = findNodeModules();
  console.info('@matimo/* packages:');

  if (!nodeModulesPath) {
    check('node_modules', false, 'node_modules not found. Run "pnpm install" or "npm install".');
    console.info('');
  } else {
    const matimoScope = path.join(nodeModulesPath, '@matimo');
    const hasScope = fs.existsSync(matimoScope);

    if (!hasScope) {
      check(
        '@matimo scope',
        false,
        'No @matimo/* packages installed. Run "matimo install slack" to get started.',
        'warn'
      );
    } else {
      const packages = fs.readdirSync(matimoScope).filter((d) => !d.startsWith('.'));

      if (packages.length === 0) {
        check(
          '@matimo scope',
          false,
          'No @matimo/* packages installed. Run "matimo install slack" to get started.',
          'warn'
        );
      } else {
        for (const pkg of packages) {
          const pkgDir = path.join(matimoScope, pkg);
          const pkgJsonPath = path.join(pkgDir, 'package.json');

          if (!fs.existsSync(pkgJsonPath)) continue;

          console.info(`  📦 @matimo/${pkg}`);

          // Check env vars for auth placeholders
          const toolsDir = path.join(pkgDir, 'tools');
          if (fs.existsSync(toolsDir)) {
            const toolDirs = fs.readdirSync(toolsDir).filter((d) => {
              return fs.existsSync(path.join(toolsDir, d, 'definition.yaml'));
            });

            const missingEnv: string[] = [];
            const seenVars = new Set<string>();

            for (const toolName of toolDirs) {
              const yamlPath = path.join(toolsDir, toolName, 'definition.yaml');
              const content = fs.readFileSync(yamlPath, 'utf-8');

              // Scan for {ENV_VAR_NAME} placeholders in headers/URLs
              const placeholderRegex = /\{(\w+)\}/g;
              let match;
              while ((match = placeholderRegex.exec(content)) !== null) {
                const name = match[1];
                if (isAuthVar(name) && !seenVars.has(name)) {
                  seenVars.add(name);
                  if (!process.env[name]) {
                    missingEnv.push(name);
                  }
                }
              }
            }

            if (missingEnv.length > 0) {
              issues.push({
                severity: 'error',
                message: `@matimo/${pkg}: missing env vars: ${missingEnv.join(', ')}`,
              });
              for (const v of missingEnv) {
                console.info(`     ❌ Missing env var: ${v}`);
              }
            } else {
              console.info(`     ✅ All required env vars are set`);
            }

            // Validate YAML schemas
            let yamlErrors = 0;
            const validateToolDefinition =
              await loadMatimoCoreExport<(t: unknown) => unknown>('validateToolDefinition');

            if (validateToolDefinition) {
              const yaml = await import('js-yaml');
              for (const toolName of toolDirs) {
                const yamlPath = path.join(toolsDir, toolName, 'definition.yaml');
                try {
                  const parsed = yaml.load(fs.readFileSync(yamlPath, 'utf-8'));
                  validateToolDefinition(parsed);
                } catch (err) {
                  yamlErrors++;
                  issues.push({
                    severity: 'error',
                    message: `@matimo/${pkg}/${toolName}: YAML invalid — ${(err as Error).message}`,
                  });
                  console.info(`     ❌ ${toolName}: invalid YAML`);
                  console.info(`        ${(err as Error).message.split('\n')[0]}`);
                }
              }
              if (yamlErrors === 0 && toolDirs.length > 0) {
                console.info(`     ✅ ${toolDirs.length} tool YAML(s) valid`);
              }
            }
          }
          console.info('');
        }
      }
    }
  }

  // ─── 3. MATIMO_APPROVAL_SECRET ─────────────────────────────────────────
  console.info('Policy / Approval:');
  const hasApprovalSecret = !!process.env.MATIMO_APPROVAL_SECRET;
  check(
    'MATIMO_APPROVAL_SECRET',
    hasApprovalSecret,
    'MATIMO_APPROVAL_SECRET is not set. Agent-created tool approvals will use a random secret (not persistent across restarts). Set it in your .env file.',
    'warn'
  );
  console.info('');

  // ─── 4. Summary ────────────────────────────────────────────────────────
  const errors = issues.filter((i) => i.severity === 'error');
  const warnings = issues.filter((i) => i.severity === 'warn');

  console.info('─'.repeat(60));
  if (errors.length === 0 && warnings.length === 0) {
    console.info('\n✅ Matimo is ready! No issues found.\n');
  } else {
    if (errors.length > 0) {
      console.info(`\n❌ ${errors.length} error(s) found — fix before using Matimo:\n`);
      errors.forEach((e, i) => console.info(`  ${i + 1}. ${e.message}`));
      console.info('');
    }
    if (warnings.length > 0) {
      console.info(`⚠️  ${warnings.length} warning(s):\n`);
      warnings.forEach((w, i) => console.info(`  ${i + 1}. ${w.message}`));
      console.info('');
    }
  }

  if (errors.length > 0) process.exit(1);
}

/** Heuristic: is this placeholder likely an auth/secret env var? */
function isAuthVar(name: string): boolean {
  const upper = name.toUpperCase();
  return (
    upper.includes('TOKEN') ||
    upper.includes('SECRET') ||
    upper.includes('KEY') ||
    upper.includes('PASSWORD') ||
    upper.includes('CREDENTIAL') ||
    upper.includes('AUTH')
  );
}

File: python/packages/cli/src/matimo_cli/commands/doctor.py
"""
``matimo doctor`` — diagnose Matimo setup.

Mirrors: packages/cli/src/commands/doctor.ts
"""
from __future__ import annotations

import os
import re
import sys
from pathlib import Path

_AUTH_PATTERNS = {"TOKEN", "SECRET", "KEY", "PASSWORD", "CREDENTIAL", "AUTH"}
_PLACEHOLDER_RE = re.compile(r"\{(\w+)\}")


def _is_auth_var(name: str) -> bool:
    upper = name.upper()
    return any(p in upper for p in _AUTH_PATTERNS)


def doctor_command() -> None:
    issues: list[dict[str, str]] = []

    def check(label: str, passed: bool, message: str, severity: str = "error") -> None:
        icon = "✅" if passed else ("❌" if severity == "error" else "⚠️ ")
        print(f"  {icon} {label}")
        if not passed:
            issues.append({"severity": severity, "message": message})
            print(f"     {message}")

    print("\n🩺 Matimo Doctor — Checking your setup…\n")

    # 1. Python version
    print("Python:")
    v = sys.version_info
    check(
        f"Python {v.major}.{v.minor}.{v.micro}",
        v >= (3, 10),
        f"Python 3.10+ required. You are running {v.major}.{v.minor}. Upgrade: https://python.org",
    )
    print()

    # 2. matimo package
    print("matimo SDK:")
    try:
        import importlib.metadata

        matimo_version = importlib.metadata.version("matimo")
        check(f"matimo v{matimo_version}", True, "")
    except importlib.metadata.PackageNotFoundError:
        check("matimo", False, 'matimo package not installed. Run "pip install matimo".')
    print()

    # 3. Installed provider packages
    print("matimo-* packages:")
    try:
        import importlib.metadata as md

        providers = [
            d
            for d in md.distributions()
            if d.metadata["Name"]
            and d.metadata["Name"].startswith("matimo-")
            and d.metadata["Name"] not in ("matimo-cli",)
        ]

        if not providers:
            check(
                "matimo-* providers",
                False,
                'No matimo-* packages installed. Run "matimo install slack" to get started.',
                "warn",
            )
        else:
            for dist in sorted(providers, key=lambda d: d.metadata["Name"]):
                pkg_name = dist.metadata["Name"]
                print(f"  📦 {pkg_name}")

                # Try to find tools directory and scan for auth placeholders
                for dist_file in dist.files or []:
                    str_path = str(dist_file)
                    if "tools/" in str_path and str_path.endswith("definition.yaml"):
                        full_path = Path(str(dist.locate_file(dist_file)))
                        if full_path.is_file():
                            content = full_path.read_text(encoding="utf-8")
                            missing = []
                            for m in _PLACEHOLDER_RE.finditer(content):
                                name = m.group(1)
                                if _is_auth_var(name) and not os.environ.get(name):
                                    missing.append(name)
                            if missing:
                                for v in missing:
                                    print(f"     ❌ Missing env var: {v}")
                                issues.append(
                                    {
                                        "severity": "error",
                                        "message": f"{pkg_name}: missing env vars: {', '.join(missing)}",
                                    }
                                )
                            else:
                                print("     ✅ All required env vars are set")
                print()
    except Exception:
        check("matimo-* scan", False, "Failed to scan installed packages.", "warn")
    print()

    # 4. MATIMO_APPROVAL_SECRET
    print("Policy / Approval:")
    has_secret = bool(os.environ.get("MATIMO_APPROVAL_SECRET"))
    check(
        "MATIMO_APPROVAL_SECRET",
        has_secret,
        "MATIMO_APPROVAL_SECRET is not set. Agent-created tool approvals will use a "
        "random secret (not persistent across restarts). Set it in your .env file.",
        "warn",
    )
    print()

    # 5. Summary
    errors = [i for i in issues if i["severity"] == "error"]
    warnings = [i for i in issues if i["severity"] == "warn"]

    print("─" * 60)
    if not errors and not warnings:
        print("\n✅ Matimo is ready! No issues found.\n")
    else:
        if errors:
            print(f"\n❌ {len(errors)} error(s) found — fix before using Matimo:\n")
            for idx, e in enumerate(errors, 1):
                print(f"  {idx}. {e['message']}")
            print()
        if warnings:
            print(f"⚠️  {len(warnings)} warning(s):\n")
            for idx, w in enumerate(warnings, 1):
                print(f"  {idx}. {w['message']}")
            print()

    if errors:
        sys.exit(1)
